### PR TITLE
Read can also accept a cancellation token

### PR DESCRIPTION
**Pull request by danielmarbach**
Split out of further addition to #231 

Cooperative cancelation is not something that is strictly only for `Task` returning APIs. 

https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads

### CODE REVIEW

File: SimpleExec/Command.cs [hunk 1]
```diff
@@ -90,22 +90,23 @@ public static async Task RunAsync(string name, string args = null, string workin
...
         {
             using (var process = new Process())
             {
                 process.StartInfo = ProcessStartInfo.Create(name, args, workingDirectory, true, windowsName, windowsArgs, configureEnvironment, createNoWindow);
 
-                var runProcess = process.RunAsync(noEcho, echoPrefix ?? DefaultPrefix.Value, CancellationToken.None);
+                var runProcess = process.RunAsync(noEcho, echoPrefix ?? DefaultPrefix.Value, cancellationToken);
                 var readOutput = process.StandardOutput.ReadToEndAsync();
 
-                Task.WaitAll(runProcess, readOutput);
```
Comment (danielmarbach) [inline SimpleExec/Command.cs:108]: Upon cancellation, WaitAll throws an AggregateException containing an inner exception as OperationCanceledException.

```
> await Task.WhenAll(Task.FromCanceled(new CancellationToken(true)));

> System.Threading.Tasks.TaskCanceledException: A task was canceled.
Task.WaitAll(Task.FromCanceled(new CancellationToken(true)));
> System.AggregateException: One or more errors occurred.
  + System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[], int, System.Threading.CancellationToken)
  + System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[], int)
  + System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[])
```

so when doing sync over async `GetAwaiter().GetResult()` is our evil friend :(

**Comment by adamralph** [comment]
@danielmarbach I'm not sure about this one yet. Is there a use case for calling `Read(..., myToken)` over `Task<string> ReadAsync(..., myToken)`?

Assuming there is, for symmetry, I'd also want to provide `Run(..., CancellationToken)`. But that probably means changing the implementation of that method to just call `RunAsync` with `GetAwaiter().GetResult()` slapped on top of it. And that leads me to wonder if both `Run` and `Read` should simply do that.

**Comment by danielmarbach** [comment]
Well the thing is when you have both sync and async flavours I believe it should have symmetry or then the sync flavour should not exist. And yes that would mean we should go one step further and have Run also support cancellation.

> But that probably means changing the implementation of that method to just call RunAsync with GetAwaiter().GetResult() slapped on top of it.  

That is a possible although dirty way to achieve it yes. From a purists standpoint if you provide sync versions they should not do sync over async and you would need to implement them redundantly to avoid nasty deadlocks in "contextual environments" but I guess for the purpose of this lib one could argue a usage of simpleexec in such a contextual env is probably an abuse of the lib anyway and since perf doesn't really matter sync over async might be OK (like it does today). Sorry if I'm stating the obvious. I'm trying to think out loud in written form because I don't have much context of simpleexec

**Comment by adamralph** [comment]
Oops, this got closed because I deleted the target branch, which was just my working copy of the branch from https://github.com/adamralph/simple-exec/pull/231.

I've [reopened](https://github.com/adamralph/simple-exec/issues/230#issuecomment-733551409) the related issue and captured this proposal there.

**Pull request closed without merging**